### ISSUE TITLE
bnd-export-maven-plugin random compile/resolution error

### ISSUE THREAD

**Issue by mrulli** [bndtools/bnd#1754]
Env: OSX Sierra; Java 1.8.0_74; Maven 3.2.3 

Project where the errors appear: https://github.com/mrulli/com.flairbit.eclipsecon.onesix

---

I slightly modified the [bnd-maven-plugin tutorial](http://enroute.osgi.org/tutorial_eval/050-start.html) adding some more dependencies. As a result I get quite pseudo-random exceptions during `mvn install`, esp. from the `example.cassandra.itest.bndrun` project.

In particular, if I add the `example.cassandra.itest.bndrun` project in the parent pom modules and I run the `mvn install` from the top level directory, the error becomes almost systematic. 

One of the stacktrace I get is the following (although also other kind errors pop up now and then):

```
[ERROR] Failed to execute goal biz.aQute.bnd:bnd-export-maven-plugin:3.4.0-SNAPSHOT:export (default) on project example.cassandra.bndrun: null: MojoExecutionException: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal biz.aQute.bnd:bnd-export-maven-plugin:3.4.0-SNAPSHOT:export (default) on project example.cassandra.bndrun: null
 at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
 at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
 at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
 at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
 at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
 at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
 at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
 at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
 at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
 at org.apache.maven.cli.MavenCli.execute(MavenCli.java:582)
 at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:214)
 at org.apache.maven.cli.MavenCli.main(MavenCli.java:158)
 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.lang.reflect.Method.invoke(Method.java:498)
 at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
 at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
 at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
 at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException
 at aQute.bnd.maven.export.plugin.ExportMojo.execute(ExportMojo.java:52)
 at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
 at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
 ... 19 more
Caused by: java.lang.NullPointerException
 at org.apache.felix.resolver.Candidates.prepare(Candidates.java:946)
 at org.apache.felix.resolver.ResolverImpl.getInitialCandidates(ResolverImpl.java:502)
 at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:387)
 at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:375)
 at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:329)
 at biz.aQute.resolve.BndResolver.resolve(BndResolver.java:29)
 at biz.aQute.resolve.ResolveProcess.resolveRequired(ResolveProcess.java:75)
 at biz.aQute.resolve.ProjectResolver.resolve(ProjectResolver.java:109)
 at biz.aQute.resolve.ProjectResolver.getRunBundles(ProjectResolver.java:126)
 at aQute.bnd.maven.export.plugin.ExportMojo.resolve(ExportMojo.java:86)
 at aQute.bnd.maven.export.plugin.ExportMojo.export(ExportMojo.java:70)
 at aQute.bnd.maven.export.plugin.ExportMojo.execute(ExportMojo.java:49)
 ... 21 more
```

As soon as I have more time to investigate this I'll try to narrow down the problem...

**Comment by j0xaf** [comment]
I had similar problems. In my case the root cause of that NPE was a proxy setup in `~/.m2/settings.xml` for `<protocol>https</https>`. Apparently bnd did not like this scheme, allowing only `http` and the like. I switched then to a dedicated `~/.bnd/settings.xml` not containing that proxy setting.

**Comment by mrulli** [comment]
I both removed the ~/.m2/settings.xml and emptied the ~/.bnd/settings.xml to get rid of some nexus settings that are not needed for the specific project that generates this issue. Unfortunately the problem endures.

**Comment by mrulli** [comment]
The NPE is generated in the `prepare()` method - `org.apache.felix.resolver.Candidates` class, within the following snippet:

```
m_candidateMap.put(r, cands.copy());
for (Capability cand : cands.getRemainingCandidates())
{
    Set<Requirement> dependents = m_dependentMap.get(cand);
    dependents.remove(origReq);
    dependents.add(r);
}
```

`org.apache.felix.resolver.Candidates.m_dependentMap.get` is invoked with key 

```	
	WrappedCapability: '[slf4j.api version=1.6.1] osgi.wiring.package; org.slf4j.impl' 
		m_cap: Provide[osgi.wiring.package]{version=1.6.1, osgi.wiring.package=org.slf4j.impl}{}
		m_host: slf4j.api version=1.6.1
```
and this nullifies `dependents` var. The only WrappedCapabilities within the m_dependentMap key set are the followings:

```
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j)(version>=1.6.1))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.spi)(version>=1.6.1))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.apache.log4j))
	 [slf4j.api version=1.6.1] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.impl)(version>=1.6.0))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.helpers)(version>=1.6.1))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.apache.log4j))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.helpers)(version>=1.6.1))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.impl)(version>=1.6.0))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.spi)(version>=1.6.1))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.impl)(version>=1.6.0))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j)(version>=1.6.1))
```
and none of them match the specified key above, at least as long as the WrappedCapability equals method is concerned.
Is there any "resolver" guru that can help on this? In my opinion a meaningful error should be thrown by the resolver instead of throwing NullPointerException. Or maybe it is just enough to skip actions on `dependents` in case null value is returned?

**Comment by mrulli** [comment]
The NPE is generated in the `prepare()` method - `org.apache.felix.resolver.Candidates` class, within the following snippet:

```
m_candidateMap.put(r, cands.copy());
for (Capability cand : cands.getRemainingCandidates())
{
    Set<Requirement> dependents = m_dependentMap.get(cand);
    dependents.remove(origReq);
    dependents.add(r);
}
```

`org.apache.felix.resolver.Candidates.m_dependentMap.get` is invoked with key 

```	
	WrappedCapability: '[slf4j.api version=1.6.1] osgi.wiring.package; org.slf4j.impl' 
		m_cap: Provide[osgi.wiring.package]{version=1.6.1, osgi.wiring.package=org.slf4j.impl}{}
		m_host: slf4j.api version=1.6.1
```
and this nullifies `dependents` var. The only WrappedCapabilities within the m_dependentMap key set are the followings:

```
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j)(version>=1.6.1))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.spi)(version>=1.6.1))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.apache.log4j))
	 [slf4j.api version=1.6.1] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.impl)(version>=1.6.0))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.helpers)(version>=1.6.1))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.apache.log4j))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.helpers)(version>=1.6.1))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.impl)(version>=1.6.0))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.spi)(version>=1.6.1))
	 [slf4j.api version=1.7.7] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j.impl)(version>=1.6.0))
	 [slf4j.api version=1.7.5] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j)(version>=1.6.1))
```
and none of them match the specified key above, at least as long as the WrappedCapability equals method is concerned.
Is there any "resolver" guru that can help on this? In my opinion a meaningful error should be thrown by the resolver instead of throwing NullPointerException. Or maybe it is just enough to skip actions on `dependents` in case null value is returned?

**Comment by bjhargrave** [comment]
You may need to file a JIRA against the Apache Felix Resolver with this information.

**Comment by mrulli** [comment]
Done: https://issues.apache.org/jira/browse/FELIX-5389

**Comment by rotty3000** [comment]
I believe this is the same issue as #1697 for which I've submitted a PR. Could you please check if that fixes the issue? It did for me. https://github.com/bndtools/bnd/pull/1759

**Comment by rotty3000** [comment]
You can test by first clearing any index files created by your repo configurations. (in the maven plugin this is as simple as deleting the target dir)

**Comment by mrulli** [comment]
This is a workaround for my specific project (to be added to itest.bndrun and com.flairbit.example.cassandra.bndrun bndrun files):

```
-runbundles.log: \
	slf4j.api; version='[1.7.7,1.7.8)',\
	slf4j.log4j12; version='[1.6.1,1.6.2)'

-runblacklist: \
	osgi.identity;filter:='(osgi.identity=slf4j.log4j12)', \
	osgi.identity;filter:='(osgi.identity=slf4j.api)'
```

But I think that also bumping the felix resolver version to 1.10.1 (thus getting the fix for https://issues.apache.org/jira/browse/FELIX-5389) could help as well. What do you think?

**Comment by bjhargrave** [comment]
Fixed by #1768
